### ISSUE TITLE
Cannot select method using fully qualified method name if it contains parameterized type information

### ISSUE THREAD

**Issue by no2chem** [junit-team/junit5#965]
Version: JUnit Jupiter 5.0.0-M5

I have a custom parameter resolver for tests that have a signature like:

```java
@Test 
void doTest(Map<String, String> parameterMap)
```

Unfortunately, the default launcher fails with this error:

```
WARNING: TestEngine with ID 'junit-jupiter' failed to discover tests
org.junit.platform.commons.JUnitException: Failed to load parameter type [java.util.Map<java.lang.String
```

It also doesn't work with one type parameter:

```java
@Test 
void doTest(Set<String> parameterMap)
```

```
WARNING: TestEngine with ID 'junit-jupiter' failed to discover tests
org.junit.platform.commons.JUnitException: Failed to load parameter type [java.util.Set<java.lang.String>]
```

If I take out the type arguments, everything behaves as expected.

Expected behavior: 

JUnit parameter resolution should work with generic types.

**Comment by sormuras** [comment]
How did you implement `supportsParameter` in your `ParameterResolver`? Do you check the type parameters available at runtime?

I tried it with the following **_quick-hack_** (don't this at home) and it works:

```java
//...
@Override
public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
  String map = "java.util.Map<java.lang.String, java.lang.String>";
  return parameterContext.getParameter().getParameterizedType().getTypeName().equals(map);
}
//...
```

**Comment by no2chem** [comment]
It seems to fail even if I do:

```
    @Override
    public boolean supportsParameter(ParameterContext parameterContext,
                                     ExtensionContext extensionContext)
            throws ParameterResolutionException {
        return true;
```

Which version does this quick hack aork on?

**Comment by sormuras** [comment]
M6

**Comment by no2chem** [comment]
I see, I'll run it against M6 once it hits central, thanks!

**Comment by sormuras** [comment]
See [MapOfStringsParameterResolver.java](https://github.com/junit-team/junit5/blob/master/junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/execution/injection/sample/MapOfStringsParameterResolver.java) and [ParameterResolverTests.java](https://github.com/junit-team/junit5/blob/master/junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/ParameterResolverTests.java#L391)

**Comment by sormuras** [comment]
Already available at
http://central.maven.org/maven2/org/junit/jupiter/junit-jupiter-engine/5.0.0-M6

**Comment by no2chem** [comment]
Great, I can confirm this is resolved in M6. Thanks for the quick response.

On Tue, Jul 18, 2017 at 4:18 PM, Christian Stein <notifications@github.com> wrote:
Already available at
http://central.maven.org/maven2/org/junit/jupiter/junit-jupiter-engine/5.0.0-M6

**Comment by sormuras** [comment]
You're welcome @no2chem

**Comment by sormuras** [comment]
Note: it works with M5, too.

```
java
-ea
-jar
  junit-platform-console-standalone\junit-platform-console-standalone-1.0.0-M5.jar
--class-path
  target\build\classes
--scan-classpath
.
+-- JUnit Jupiter [OK]
| +-- Tests [OK]
| | +-- testMapOfStrings(Map) [OK]
(...)
```

**Comment by sormuras** [comment]
Note 2: use a library like https://github.com/jhalterman/typetools to retrieve runtime type information

**Comment by sbrannen** [comment]
Reopening this issue since I have been able to reproduce the exception on 5.0 M6.

**Comment by sbrannen** [comment]
**Analysis**:

@sormuras has done a great job of demonstrating how one should implement a `ParameterResolver` for method parameters with parameterized type information.

However, the exception mentioned in this issue's description is actually not about parameter resolution. Rather, it is about selecting a method by its _fully qualified method name_ (FQMN) when the parameter names include parameterized type information.

For example, if the parameterized type information is omitted then the following FQMN can successfully be used to select a method: `org.example.ParameterizedTypeTests#testMapOfStrings(java.util.Map)`.

However, if parameterized type information is included in the FQMN as follows:

`org.example.ParameterizedTypeTests#testMapOfStrings(java.util.Map<java.lang.String, java.lang.String>)`

... then we see the following exception:

```
org.junit.platform.commons.JUnitException: Failed to load parameter type [java.util.Map<java.lang.String]
 ```

`, java.lang.String>` is omitted due to the fact that the regular expression used in `ReflectionUtils.loadMethod(String)` _splits_ the parameter arguments by commas.

**Comment by sbrannen** [comment]
@no2chem, can you please tell us how you selected the method to execute?

Were you using the `Launcher` API directly?

Were you using an IDE?

Were you using Maven or Gradle?

**Comment by sbrannen** [comment]
FYI: I have updated the title to address the underlying issue.

**Comment by sbrannen** [comment]
Please note that the JavaDoc for `org.junit.platform.engine.discovery.DiscoverySelectors.selectMethod(String)` describes the supported _parameter type list_ as follows.

> The _parameter type list_ is a comma-separated list of primitive names or fully qualified class names for the types of parameters accepted by the method.

Chapter [6.7. Fully Qualified Names and Canonical Names](http://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.7) of the Java Language Specification states the following.

> The fully qualified name of a top level class or top level interface that is declared in a named package consists of the fully qualified name of the package, followed by ".", followed by the simple name of the class or interface. 

Consequently, `java.util.Map<java.lang.String, java.lang.String>` is not a _fully qualified class name_ since it contains generic type information.

In summary, the manner in which the method was selected was incorrect; however, we could potentially add a check for the presence of `<` or `>` in a FQMN and log a warning or throw a `PreconditionViolationException`.

@junit-team/junit-lambda, thoughts?

**Comment by no2chem** [comment]
Interesting. After updating to M6 the problem has disappeared. I was using IntelliJ, and I think the problem was occurring in Surefire under a Maven build as well (when selecting a test class using the -Dtest= parameter.

**Comment by sbrannen** [comment]
Based on the feedback provided by @no2chem in #966 as well as the input here, it is clear that it is IntelliJ that is invoking `DiscoverySelectors.selectMethod(String)` with generic type information for parameters instead of _fully qualified class names_ for parameter types.

@akozlova, can you please confirm what IntelliJ passes to `DiscoverySelectors.selectMethod(String)` when a user selects an individual method to execute on the JUnit Platform?

**Comment by sbrannen** [comment]
> Interesting. After updating to M6 the problem has disappeared. I was using IntelliJ, and I think the problem was occurring in Surefire under a Maven build as well (when selecting a test class using the -Dtest= parameter.

Thanks for the feedback, @no2chem!

However, since you raised #966, I can only assume that the issue you originally raised here still exists _if you select a single method to execute_.

Is that assumption correct?

**Comment by no2chem** [comment]
Yes, that is correct. M6 appears to have solved the problem if I launch the entire test class, however  running individual methods still appears to present the same issue.

**Comment by sbrannen** [comment]
> Yes, that is correct. M6 appears to have solved the problem if I launch the entire test class, however running individual methods still appears to present the same issue.

OK. Thanks for the prompt feedback! 👍 

We'll try to sort this out with @akozlova from IntelliJ.

Cheers

**Comment by akozlova** [comment]
Intellij start ```testMe(java.util.Map)``` for ```testMe(Map<String, String> m)```

**Comment by no2chem** [comment]
@akozlova I see, it's possible this is fixed in 2017.2, I am running 2017.1.5. I'll do an update now.

**Comment by sbrannen** [comment]
> Intellij start ```testMe(java.util.Map)``` for ```testMe(Map<String, String> m)```

OK. That looks good.

Thanks for confirming, @akozlova!

I'm guessing/hoping it's just an issue in a previous version of IntelliJ like @no2chem just proposed.

**Comment by akozlova** [comment]
Yes, the signature was not erased in 2017.1. Please check 2017.2 (just released). 

Please note that IntelliJ still bundles M4, M5 will be bundled in 2017.2.1 EAP (coming this week, I think), M6 need to be checked first and then probably would appear in 2017.2.2.

**Comment by sbrannen** [comment]
Hi @akozlova,

This is the right place to look in the IntelliJ code base -- right?

http://git.jetbrains.org/?p=idea/community.git;a=tree;f=plugins/junit5_rt/src/com/intellij/junit5;h=e7b91d7488a52b67b2834a0228296fc1ee30078d;hb=HEAD

I see that in `JUnit5TestRunnerUtil` you have the following method:

```java
  protected static DiscoverySelector createSelector(String line) {
    if (line.contains(",")) {
      return DiscoverySelectors.selectMethod(line.replaceFirst(",", "#"));
    }
    else {
      return DiscoverySelectors.selectClass(line);
    }
  }
```

But... I don't know what the content of the `line` variable looks like.

How can I see what types of values are passed into that (for my own future reference... in order to better help diagnose such issues)?

Cheers,

Sam

**Comment by akozlova** [comment]
You asked it!

https://github.com/JetBrains/intellij-community/blob/b59ffe20e43926797c3bad8e7034c2323749982d/plugins/junit/src/com/intellij/execution/junit/JUnitConfiguration.java#L631

you find the part which goes to the clients VM but the selector is prepared in IDEA's VM.

**Comment by no2chem** [comment]
@akozlova Great, 2017.2 works great on generics and correctly erases the type signature. However, it doesn't quite work on type parameters, as it seems to delete the type parameters instead of resolving. I can provide a test case.

**Comment by no2chem** [comment]
Type parameter example:

```java
public interface MapContract<M extends Map<K,V>, K, V> extends TestableMap<M, K, V> {
    @InterfaceTest
    default void canPut(M map, K k1, V v1) {
        map.put(k1, v1);

        assertThat(map)
                .containsExactly(MapEntry.entry(k1, v1));
    }
}

public class HashMapIntegerTest implements MapContract<HashMap, Integer, Integer> {
    @Override
    public HashMap<Integer, Integer> createMap() {
        return new HashMap<>();
    }

    AtomicInteger keyCounter = new AtomicInteger();
    @Override
    public Integer createKey() {
        return keyCounter.getAndIncrement();
    }

    AtomicInteger valueCounter = new AtomicInteger();
    @Override
    public Integer createValue() {
        return valueCounter.getAndIncrement();
    }
}

@TestableInterface
public interface TestableMap<M extends Map<K,V>, K, V> {
    M createMap();
    K createKey();
    V createValue();
}
```

**Comment by akozlova** [comment]
It leaves type parameter as is, though could be erased as well. Thanks

**Comment by no2chem** [comment]
@sbrannen, from the exception above, is discovery not able to find testPut because it's a member of an interface rather than the class, or because the parameters were not supplied by IntelliJ?

**Comment by akozlova** [comment]
You can always pass the method signature manually to check how the junit engine will work with it,

**Comment by sbrannen** [comment]
> You asked it!

Indeed I did. 😱 

Thanks, @akozlova

**Comment by sbrannen** [comment]
> You can always pass the method signature manually to check how the junit engine will work with it, BTW surely IntelliJ has an issue tracker and the link is fine.

Exactly!

@no2chem, if you need help figuring out how to do that, just ask.

**Comment by sbrannen** [comment]
@no2chem and @akozlova,

Did either of you create an issue in the IntelliJ issue tracker for the "type parameters are not erased in method signatures"?

**Comment by sbrannen** [comment]
I ask, because I'd like to close this issue, but I want to make sure it gets handle on the IntelliJ side.

**Comment by sbrannen** [comment]
_waiting for feedack_

**Comment by akozlova** [comment]
I've commit a fix without an issue, sorry

**Comment by sbrannen** [comment]
ok. no worries.

The "fix" is what we're really after anyway. 😉 

Closing this issue now...

Cheers!

**Comment by sbrannen** [comment]
@no2chem,

> This looks like maybe it's an IntelliJ rather than an JUnit issue,

Actually, that might be an issue for both projects.

Thanks to your questioning, I've created a _failing_ test against `master` where a parameterized default method cannot be found.

So I'll open a new issue to address that.

Cheers!

**Comment by sbrannen** [comment]
FYI: the new issue is #969.
